### PR TITLE
advancing to scalars gen2

### PR DESCRIPTION
**Pull request by YoelShoshan**

### CODE REVIEW

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 1]
```diff
@@ -147,10 +144,12 @@ def prepare_info_for_model_step(
...
-        scalars_indices = []
-        scalars_values = []
-        scalars_masked_indices = []
-        prev_index_end = -1
+        ## both `all_scalars_values` and `all_scalars_valid_mask` will contain torch tensors, which will be concatanated in the end of this function
+        all_scalars_values = (
+            []
+        )  # one scalar for every element, `scalar_default_unfound_value` is used for elements that aren't scalars
+        all_scalars_valid_mask = []  # for each element, whether it's a scalar or not
+        scalar_default_unfound_value = -1000.0
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:152]: What does it mean?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 2]
```diff
@@ -47,7 +45,7 @@ class InjectorToModularTokenizerLib:
     @staticmethod
     def build_placeholder_meta_tokenization(
         *,
-        sequence: Union[str, list, tuple],
+        sequence: str,
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:48]: Why did you change it if you still support those types?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 3]
```diff
@@ -91,19 +88,18 @@ def build_placeholder_meta_tokenization(
             if tokenizer_type.startswith("SCALARS_"):
                 with_placeholders.append(
                     "<@TOKENIZER-TYPE=AA>"
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:90]: The general solution can't assume there is AA subtokenizer.
Maybe we need a default empty sub-tokenizer? Maybe SCALARS can be an empty sub-tokenizer?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 4]
```diff
@@ -91,19 +88,18 @@ def build_placeholder_meta_tokenization(
...
-                )  # won't use AA tokens, just an arbitrary one to be able to use a token like <SCALAR>
+                )  # AA tokenizer selection is arbitrary, we only take the special token <SCALAR> from it
 
-                if (
-                    tokenizer_type == "SCALARS_LITERALS"
-                ):  # note: masking is only supported in literals (not in "from dict")
+                if tokenizer_type == "SCALARS_LITERALS":
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:93]: Remind me, can we put mask in scalar literals?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 5]
```diff
@@ -147,10 +144,12 @@ def prepare_info_for_model_step(
 
 
         """
-        scalars_indices = []
-        scalars_values = []
-        scalars_masked_indices = []
-        prev_index_end = -1
+        ## both `all_scalars_values` and `all_scalars_valid_mask` will contain torch tensors, which will be concatanated in the end of this function
+        all_scalars_values = (
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:148]: all_scalar_values = [] ?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 6]
```diff
@@ -91,19 +88,18 @@ def build_placeholder_meta_tokenization(
...
+                if tokenizer_type == "SCALARS_LITERALS":
                     values = subseq.split(",")
-                    # seq = "<SCALAR>" * len(values)
-                    seq = "".join(
-                        [
-                            "<MASKED_SCALAR>" if x == "<MASK>" else "<SCALAR>"
-                            for x in values
-                        ]
-                    )
+                    # validate that all values can be converted to fload
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:95]: typo: float

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 7]
```diff
@@ -209,24 +196,47 @@ def prepare_info_for_model_step(
...
+                )
+
+        all_scalars_values = torch.concat(all_scalars_values)
+        all_scalars_valid_mask = torch.concat(all_scalars_valid_mask)
+
+        assert all_scalars_values.shape == all_scalars_valid_mask.shape
+
+        # pad if needed
+        full_query_len = len(token_ids)
+        if full_query_len > all_scalars_values.shape[0]:
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:218]: Can't it happend?

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 8]
```diff
@@ -130,6 +126,7 @@ def prepare_info_for_model_step(
         *,
```
Comment (mosheraboh) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:126]: maybe build_scalars be a better name here?

**Review by mosheraboh — APPROVED** [review]
Looks great!
Few comments inline/

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 1 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:152]: since we now keep the scalars values and mask at the size of the entire sequence, this is the default value for places that don't actually have a scalar value.
I chose -1000 and not something like 0 to make sure it pops up easily if there are mistakes down the road

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 3 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:90]: interesting point.
SCALARS is currently fully programmatic and does not rely on any dictionary, so I would rather not mix it.
Probably better to have "base" that gets automatically generated and supported , as the modular tokenizer already knows how to handle special tokens

maybe "Base" or "SpecialTokensBase" or something

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 2 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:48]: I just copied over the code, so probably you already fixed this and it got overriden.
fixed it now to contain the union.

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 5 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:148]: I think that static code auto "fixed" this, due to the comment length
(I didn't write it like this)

Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:148]: I moved the comments to one line above the code, so it won't do that.

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 6 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:95]: thanks!
fixed

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 7 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:218]: Do you mean "can it happen" ? 
If so, then yes, almost always.

The main code logic (before it) iterates over each sub part (with specific sub tokenizer) so it does not contain the padding.

I can explain more if it isn't clear

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py (hunk 8 above)
Comment (YoelShoshan) [reply fuse/data/tokenizers/modular_tokenizer/inject_utils.py:126]: np, renamed

File: fuse/data/tokenizers/modular_tokenizer/inject_utils.py [hunk 9]
```diff
@@ -209,24 +197,52 @@ def prepare_info_for_model_step(
...
+                [
+                    all_scalars_valid_mask,
+                    torch.full(
+                        (pad_len,), fill_value=False, dtype=all_scalars_valid_mask.dtype
+                    ),
+                ]
             )
-        else:
-            scalars_masked_indices = None
+        elif full_query_len > all_scalars_values.shape[0]:
```
Comment (YoelShoshan) [inline fuse/data/tokenizers/modular_tokenizer/inject_utils.py:239]: @mosheraboh see here, related to what we talked about.

I'll try to add more unit tests with interesting cases by the end of this week

**Review by mosheraboh — APPROVED** [review]
